### PR TITLE
Add final operation method to help PCEX backend

### PR DESCRIPTION
**Pull request by corviday**
This branch adds a simple method that will allow the PCEX backend to use this package for its dataset-filtering needs rather than munging cell methods itself.

I suspect @rod-glover will have suggestions on more elegant ways to achieve the same result; this PR is more a discussion and a description of the problem space than anything else.

The PCEX backend needs to use this cell method parser to distinguish among three types of dataset, each of which can be requested separately via its API via a parameter. The three types of dataset represent different climatological statistics.

**Type 1: Standard Deviation data (parameter: standard_deviation)**
These are climatological standard deviations. They have correct cell methods. At present, the following cell methods are in this category:
* `time: mean time: standard_deviation over days`
* `time: maximum time: standard_deviation over days`
* `time: minimum time: standard_deviation over days`
though of course we may add other standard deviation data someday.

**Type 2: percentile data (parameter: percentile)**
These are percentiles over ensembles of models. This data is not yet in the database, but will have correct cell methods when it is. Their cell_methods are expected to be
* `time: mean within days time: max over days time: mean over days models: percentile[N]` (for various N)
though of course we may add other percentile data someday.

**Type 3: Everything else (parameter: mean)**
All other files available to PCEX are climatological means<sup>1</sup>.  Unfortunately, many of these files have unreliable cell_methods; at this point in time we cannot rely on parsing their cell methods to determine they are climatological means, only the process of elimination.
* `[something] time: mean over days` (made by `generate_climos`, generally correct-ish)
* `[something] time:mean over days models: mean` (PCIC12 data, occasionally, notable because it has an operation AFTER the climatological averaging but still counts as a climatological mean)
* `[something unparsable]`
* `[something parsable but wrong]`

The easiest way to distinguish these types of files given the unreliability of cell_methods for Type 3 is to check the operation of the last method. It is either `standard_deviation` (type 1), `percentile[N]` (type 2) `mean` (type 3), anything else (type 3), or the cell method is unparsable (also type 3). Accordingly, I have added a function that just returns the operation string of the final method for the PCEX backend's use.

I'm open to other approaches. 

We could make a predicate for each type, in the style of the functions in semantics.py, but depending on how we wrote them, we might have to update the predicates if we got more files added to that type. (Plus we'd maybe want several different predicates for type three to cover all cases?)

Alternately, we could have the backend do the parsing if we preferred. Currently it is using a fairly fragile regex, but it could use the tools this package provides instead. 

There are probably other options.

Thoughts on the best approach, @rod-glover ?

-------------------

<sup>1</sup> With the exception of the physical geography files, which are not climatological statistics, used by the `/watershed` API. Nobody should be using any other API with them so they don't really enter into this conversations. They are distinguishable by having all their cell methods over `area:` instead of time. They have no time component.

### CODE REVIEW

File: cf_cell_methods/pcex_helpers.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#     * models: percentile[n] (hydrology percentile data - used for context)
+# Since PCEX treats means over model ensembles and means over time identically, 
+# all the backend needs to know is the last cell method - "mean", "standard_deviation",
+# "percentile", or other / None. 
+def final_operation(cell_method):
+    try:
+        cm = parse(cell_method)
+        method_str = cm[-1].method.name
+        #strip percentile bonus information if present
+        return method_str.split("[")[0]
```
Comment (rod-glover) [inline cf_cell_methods/pcex_helpers.py:27]: You shouldn't have to do this! The _intention_, at least, is that `cm.method` is an object of type [`Method`](https://github.com/pacificclimate/cf-cell-methods/blob/add_final_method/cf_cell_methods/representation.py#L120). `Method` has two attributes, `name` and `params`. So you ought to just be able to compare `cm.method` as is, no splitting required. This probably speaks to deficient documentation, unless there's a bug in the implementation, but [I don't think so](https://github.com/pacificclimate/cf-cell-methods/blob/add_final_method/tests/test_parser.py#L15-L18).

File: cf_cell_methods/pcex_helpers.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#     * time: mean over days (a climatological mean - user can view this)
+#     * time: standard_deviation over days (a climatological SD - used for context)
+#     * models: mean (PCIC12 climatologies - user can view this)
+#     * models: percentile[n] (hydrology percentile data - used for context)
+# Since PCEX treats means over model ensembles and means over time identically, 
+# all the backend needs to know is the last cell method - "mean", "standard_deviation",
+# "percentile", or other / None. 
+def final_operation(cell_method):
+    try:
+        cm = parse(cell_method)
```
Comment (rod-glover) [inline cf_cell_methods/pcex_helpers.py:24]: Minor point, but with the already somewhat confusing terminology imposed on us by CF, better to be as careful as we can: The term for the entire string is "cell methods". In some of our code and in the database, the name is transcribed in the singular, `cell_method`, but it should be `cell_methods`. A single cell method is one component of a cell methods string. Sorry to be pedantic.

File: cf_cell_methods/pcex_helpers.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# The functions in this file are helpers for the PCEX backend.
```
Comment (rod-glover) [inline cf_cell_methods/pcex_helpers.py:1]: Should be a docstring.

File: cf_cell_methods/pcex_helpers.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+# additional context to the dataset selected by the user. They are handled
+# differently by the PCEX backend.
+# Some example final cell methods:
+#     * time: mean over days (a climatological mean - user can view this)
+#     * time: standard_deviation over days (a climatological SD - used for context)
+#     * models: mean (PCIC12 climatologies - user can view this)
+#     * models: percentile[n] (hydrology percentile data - used for context)
+# Since PCEX treats means over model ensembles and means over time identically, 
+# all the backend needs to know is the last cell method - "mean", "standard_deviation",
+# "percentile", or other / None.
```
Comment (rod-glover) [inline cf_cell_methods/pcex_helpers.py:21]: This should be a docstring for `final_operation`.

**Review by rod-glover — CHANGES_REQUESTED** [review]
Comments:

- It seems reasonable to put helper methods here if they are useful to many clients. But if this is just for PCEX, then it might make more sense to put them in that codebase. Then you don't have to do a new release of `cf-cell-methods` to fix a bug or add a new helper. The trick is distinguishing general utility from specific application. I think this might fall more into the latter category.
- The general approach you outline (examining the final operation) seems not unreasonable, but I wonder if it is robust enough. You could compare for the whole sequence of cell methods in each case, using something like, for case 1, this
   ```
   len(cell_methods) == 2 and cm[0].name == "time" and cm[1].method.name == "standard_deviation"
   ```
   or even, more specifically, this
   ```
   len(cell_methods) == 2 and cm[0].name == "time" and cm[1] == CellMethod("time", Method("standard_deviation", None), over="days")
   ```
   or
   ```
   len(cell_methods) == 2 and cm[0].name == "time" and cm[1] == parse("standard_deviation over days")[0]
   ```
   All the representation classes know how to compare themselves, so the equality comparisons between representation objects above can be used as an alternative to picking out all the attributes by hand. It would be better if I had provided a single-cell-method parser so that one can avoid the clunky `parse(...)[0]`, but I didn't (yet).

**Comment by corviday** [comment]
>  The general approach you outline (examining the final operation) seems not unreasonable, but I wonder if it is robust enough. You could compare for the whole sequence of cell methods in each case, using something like, for case 1, this

Hm. My thinking on only examining the _last_ cell method was that while currently we only have one set of standard deviation climatologies (anusplin tasmax, tasmin, and pr for plan2adapt) and only one set of percentile climatologies (hydrology's new streamflow extremes), it seems very plausible we will have more someday. 

Those hypothetical new datasets could have completely different numbers and types of methods before the percentile or standard deviation method, but they would have the same final method. So I didn't want to specify what happened before the final statistical operation, so that if we ever got more datasets that belonged in these categories, we could drop them in without changing the code. Does that seem reasonable? Or is it preferable to characterize the whole sequence of the cell_methods string in more detail now, and just change it if we get more percentile or standard deviation data?

> But if this is just for PCEX, then it might make more sense to put them in that codebase. 
I will delete this PR when we are done discussing and do that instead.

**Comment by rod-glover** [comment]
> Hm. My thinking on only examining the _last_ cell method was that while currently we only have one set of standard deviation climatologies (anusplin tasmax, tasmin, and pr for plan2adapt) and only one set of percentile climatologies (hydrology's new streamflow extremes), it seems very plausible we will have more someday.
> 
> Those hypothetical new datasets could have completely different numbers and types of methods before the percentile or standard deviation method, but they would have the same final method. So I didn't want to specify what happened before the final statistical operation, so that if we ever got more datasets that belonged in these categories, we could drop them in without changing the code. Does that seem reasonable? Or is it preferable to characterize the whole sequence of the cell_methods string in more detail now, and just change it if we get more percentile or standard deviation data?

That sounds quite reasonable. I still feel slightly cautious about casting too big a net, but if you feel confident right now about this then :+1: 

> > But if this is just for PCEX, then it might make more sense to put them in that codebase.

:+1:

**Comment by corviday** [comment]
We decided to implement this i[n the backend](https://github.com/pacificclimate/climate-explorer-backend/pull/176) instead.

**Comment by corviday** [comment]
We decided to implement this [in the backend](https://github.com/pacificclimate/climate-explorer-backend/pull/176) instead.

**Pull request closed without merging**